tendermint/src/proposal.rs

```diff
@@ -0,0 +1,267 @@
+        }
+    }
+}
+
+impl Proposal {
+    /// Create signable bytes from Proposal.
+    pub fn to_signable_bytes<B>(
+        &self,
+        chain_id: ChainId,
+        sign_bytes: &mut B,
+    ) -> Result<bool, DomainTypeError>
+    where
+        B: BufMut,
+    {
+        CanonicalProposal::new(self.clone(), chain_id).encode_length_delimited(sign_bytes)?;
+        Ok(true)
+    }
+
+    /// Create signable vector from Proposal.
+    pub fn to_signable_vec(&self, chain_id: ChainId) -> Result<Vec<u8>, DomainTypeError> {
+        CanonicalProposal::new(self.clone(), chain_id).encode_length_delimited_vec()
+    }
+
+    /// Consensus state from this proposal - This doesn't seem to be used anywhere.
+    #[deprecated(
+        since = "0.17.0",
+        note = "This seems unnecessary, please raise it to the team, if you need it."
+    )]
+    pub fn consensus_state(&self) -> State {
+        State {
+            height: self.height,
+            round: self.round,
+            step: 3,
+            block_id: self.block_id,
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::block::parts::Header;
+    use crate::block::Id as BlockId;
+    use crate::block::{Height, Round};
+    use crate::chain::Id as ChainId;
+    use crate::hash::{Algorithm, Hash};
+    use crate::proposal::SignProposalRequest;
+    use crate::signature::{Ed25519Signature, ED25519_SIGNATURE_SIZE};
+    use crate::{proposal::Type, Proposal, Signature};
+    use chrono::{DateTime, Utc};
+    use std::convert::TryFrom;
+    use std::str::FromStr;
+    use tendermint_proto::DomainType;
+
+    #[test]
+    fn test_serialization() {
+        let dt = "2018-02-11T07:09:22.765Z".parse::<DateTime<Utc>>().unwrap();
+        let proposal = Proposal {
+            msg_type: Type::Proposal,
+            height: Height::try_from(12345_i64).unwrap(),
+            round: Round::try_from(23456).unwrap(),
+            pol_round: None,
+            block_id: Some(BlockId {
+                hash: Hash::from_hex_upper(
+                    Algorithm::Sha256,
+                    "DEADBEEFDEADBEEFBAFBAFBAFBAFBAFADEADBEEFDEADBEEFBAFBAFBAFBAFBAFA",
+                )
+                .unwrap(),
+                parts: Some(Header {
+                    total: 65535,
+                    hash: Hash::from_hex_upper(
+                        Algorithm::Sha256,
+                        "0022446688AACCEE1133557799BBDDFF0022446688AACCEE1133557799BBDDFF",
+                    )
+                    .unwrap(),
+                }),
+            }),
+            timestamp: Some(dt.into()),
+            signature: Signature::Ed25519(Ed25519Signature::new([0; ED25519_SIGNATURE_SIZE])),
+        };
+
+        let mut got = vec![];
+
+        let request = SignProposalRequest {
+            proposal,
+            chain_id: ChainId::from_str("test_chain_id").unwrap(),
+        };
+
+        let _have = request.to_signable_bytes(&mut got);
+
+        // the following vector is generated via:
```

Comment:
Such a great idea to include the code which was used to generate those vectors :)

Reply:
Thanks! It's not my idea, but I extended on it. I'll try to keep up the habit for easier upgrades in the future.